exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1456 @@
+   *
+   * @param upstream the status code from upstream: either OK or OK_NEW_SCHEMA
+   */
+
+  private void setupSchema(IterOutcome upstream)  {
+
+    // First batch: we won't have a schema.
+
+    if (schema == null) {
+      schema = incoming.getSchema();
+
+    // Subsequent batches, nothing to do if same schema.
+
+    } else if (upstream == IterOutcome.OK) {
+      return;
+
+    // Only change in the case that the schema truly changes. Artificial schema changes are ignored.
+
+    } else if (incoming.getSchema().equals(schema)) {
+      return;
+    } else if (unionTypeEnabled) {
+        schema = SchemaUtil.mergeSchemas(schema, incoming.getSchema());
+
+        // New schema: must generate a new sorter and copier.
+
+        opCodeGen.setSchema(schema);
+    } else {
+      throw UserException.unsupportedError()
+            .message("Schema changes not supported in External Sort. Please enable Union type.")
+            .build(logger);
+    }
+
+    // Coerce all existing batches to the new schema.
+
+    for (BatchGroup b : bufferedBatches) {
+//      System.out.println("Before: " + allocator.getAllocatedMemory()); // Debug only
+      b.setSchema(schema);
+//      System.out.println("After: " + allocator.getAllocatedMemory()); // Debug only
+    }
+    for (BatchGroup b : spilledRuns) {
+      b.setSchema(schema);
+    }
+  }
+
+  /**
+   * Convert an incoming batch into the agree-upon format. (Also seems to
+   * make a persistent shallow copy of the batch saved until we are ready
+   * to sort or spill.)
+   *
+   * @return the converted batch, or null if the incoming batch is empty
+   */
+
+  private VectorContainer convertBatch() {
+
+    // Must accept the batch even if no records. Then clear
+    // the vectors to release memory since we won't do any
+    // further processing with the empty batch.
+
+    VectorContainer convertedBatch = SchemaUtil.coerceContainer(incoming, schema, oContext);
+    if (incoming.getRecordCount() == 0) {
+      for (VectorWrapper<?> w : convertedBatch) {
+        w.clear();
+      }
+      return null;
+    }
+    return convertedBatch;
+  }
+
+  private SelectionVector2 makeSelectionVector() {
+    if (incoming.getSchema().getSelectionVectorMode() == BatchSchema.SelectionVectorMode.TWO_BYTE) {
+      return incoming.getSelectionVector2().clone();
+    } else {
+      return newSV2();
+    }
+  }
+
+  /**
+   * Process the converted incoming batch by adding it to the in-memory store
+   * of data, or spilling data to disk when necessary.
+   */
+
+  @SuppressWarnings("resource")
+  private void processBatch() {
+
+    // Skip empty batches (such as the first one.)
+
+    if (incoming.getRecordCount() == 0) {
+      return;
+    }
+
+    // Determine actual sizes of the incoming batch before taking
+    // ownership. Allows us to figure out if we need to spill first,
+    // to avoid overflowing memory simply due to ownership transfer.
+
+    RecordBatchSizer sizer = analyzeIncomingBatch();
```

Comment:
In an ideal world, each batch would know its size and the code here would simply query that number. Yes, we do check on each batch to detect the case where, say, the first batch happened to have short columns and later batches have longer columns. We need this because the "copier" works only with row counts. To ensure we don't overflow memory, we need an accurate row width estimate, hence the checks.

Further, it turns out that Parquet (only some files) create strangely-sized batches, so simply taking the batch size / row count produces non-sensical numbers (rows of width 28 K when they are really 500 bytes, say.)

Of course, the best solution would be if the "copier" worked on actual memory usage: copy data until the outgoing batch reaches such-and-so size. Then the size of the average row would be moot and this whole chunk of code could be removed. That is left as an exercise for later.